core/src/main/java/org/apache/brooklyn/core/objs/BasicSpecParameter.java

```diff
@@ -384,5 +387,73 @@ public static WeightedParameter getFieldConfig(ConfigKey<?> config, Field config
         }
     }
 
+    /**
+     * Returns true if the {@link SpecParameter parameter name} is the same as
+     * that on the specified paramater.
+     */
+    protected static Predicate<SpecParameter<?>> sameName(final SpecParameter<?> param) {
+        return new Predicate<SpecParameter<?>>() {
+            @Override
+            public boolean apply(SpecParameter<?> input) {
+                return input.getConfigKey().getName().equals(param.getConfigKey().getName());
+            }
+        };
+    }
+
+    /**
+     * Returns true if the {@link SpecParameter#getLabel() label} is the same as
+     * the specified string.
+     */
+    public static Predicate<SpecParameter<?>> labelEqualTo(final String label) {
+        return new Predicate<SpecParameter<?>>() {
+            @Override
+            public boolean apply(SpecParameter<?> input) {
+                return input.getLabel().equals(label);
+            }
+        };
+    }
+
+    /**
+     * Returns true if the {@link ConfigKey#getName() config key name} is the same
+     * as the specified string.
+     */
+    public static Predicate<SpecParameter<?>> nameEqualTo(final String name) {
+        return new Predicate<SpecParameter<?>>() {
+            @Override
+            public boolean apply(SpecParameter<?> input) {
+                return input.getConfigKey().getName().equals(name);
+            }
+        };
+    }
+
+    /**
+     * Adds the given list of {@link SpecParameter parameters} to the provided
+     * {@link AbstractBrooklynObjectSpec spec}. Inherits parent parameters except
+     * where there is an existing {@link ConfigKey config key}.
+     * <p>
+     * Implemented by explicitly replacing the existing parameters with an
+     * updated list.
+     *
+     * @see EntitySpec#parameters(List)
+     */
+    public static void addParameters(AbstractBrooklynObjectSpec<?, ?> spec, List<? extends SpecParameter<?>> explicitParams, BrooklynClassLoadingContext loader) {
+        if (explicitParams.size() > 0) {
+            List<SpecParameter<?>> currentParams = MutableList.copyOf(spec.getParameters());
+            Map<ConfigKey<?>, ?> currentConfig = spec.getConfig();
+            for (final SpecParameter<?> param : explicitParams) {
+                Optional<ConfigKey<?>> configExists = Iterables.tryFind(currentConfig.keySet(), ConfigPredicates.nameEqualTo(param.getConfigKey().getName()));
+                if (!configExists.isPresent()) {
+                    Optional<SpecParameter<?>> paramExists = Iterables.tryFind(currentParams, sameName(param));
+                    if (paramExists.isPresent()) {
+                        currentParams.remove(paramExists.get());
```

Comment:
OK, I think that might work, if I remove the code that deals with `ConfigKey` then things will be OK, because the actual config will still override a parameter with a default, no matter where in the hierarchy it is set. I'll add a test case for the inherited config case. I assume you mean something like this:

```YAML
- type: entity
  id: parent
  brooklyn.config:
    key: value
  brooklyn.children:
    - type: entity
      id: child
      brooklyn.parameters:
        - name: key
          default: other
```

So, the value of `$brooklyn:entity("child").config("key")` will be `value` because it inherits from the parent entity, yes?